pravega/pravega
# Issue 4452: (SegmentStore) Fixed Read Index Append Concurrency bug

**Change log description**  
- Fixed a bug in `StreamSegmentReadIndex` where it would have been possible for an append data to get lost if the `append` method was invoked concurrently with a cache eviction (`updateGenerations`).
 
**Purpose of the change**  
Fixes #4452.

**What the code does**  
- See parent issue comment for reproduction scenario (https://github.com/pravega/pravega/issues/4452#issuecomment-566168592)
- If a cache eviction occurs during a call to `append`, it would have been possible to append to a `ReadIndexEntry` that was just evicted or about to be. This would have led to temporary data unavailability and an eventual segment container restart (which would have addressed the issue).
- The bug was caused because even though we updated the Read Index Structure under its lock, we were modifying the last `ReadIndexEntry` to have an additional length. This should have been done under the lock.
    - It would be preferable to not invoke `CacheStorage.append` under this lock, however if we update the `ReadIndexEntry`'s length prior to actually writing the data, we risk serving garbage data to a concurrent reader (which would have been worse than the current situation). As such, we have no choice but to do the whole append under this lock. This is not too bad, given that the append itself is doing a memory copy of up to 4KB.
- This was introduced with #4074.

**How to verify it**  
All tests must pass. New unit test was added to verify that appending and cache eviction cannot run concurrently.
